File: src/main/java/carnage/actionLogger/listeners/ContainerListeners.java
package carnage.actionLogger.listeners;

import carnage.actionLogger.models.ContainerTransaction;
import carnage.actionLogger.queue.LogQueue;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.*;
import org.bukkit.event.*;
import org.bukkit.event.inventory.*;
import org.bukkit.inventory.*;

public class ContainerListeners implements Listener {
    private final LogQueue queue;
    public ContainerListeners(LogQueue queue) { this.queue = queue; }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onClick(InventoryClickEvent e) {
        if (!(e.getWhoClicked() instanceof org.bukkit.entity.Player p)) return;
        var clicked = e.getClickedInventory();
        if (clicked == null || !isContainer(clicked.getType())) return;
        var loc = getContainerLocation(clicked);
        if (loc == null) return;

        var current = e.getCurrentItem();
        var cursor = e.getCursor();
        var container = clicked.getType().name();

        switch (e.getAction()) {
            case PICKUP_ALL, PICKUP_SOME, PICKUP_HALF, PICKUP_ONE -> {
                if (current != null && current.getType() != Material.AIR) {
                    int amt = switch (e.getAction()) {
                        case PICKUP_HALF -> (current.getAmount() + 1) / 2;
                        case PICKUP_ONE -> 1;
                        default -> current.getAmount();
                    };
                    queue.addContainerTransaction(new ContainerTransaction(p.getName(), (byte)0, container,
                            current.getType().name(), amt, loc.getWorld().getName(),
                            loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
                }
            }
            case PLACE_ALL, PLACE_SOME, PLACE_ONE -> {
                if (cursor != null && cursor.getType() != Material.AIR) {
                    int amt = e.getAction() == InventoryAction.PLACE_ONE ? 1 : cursor.getAmount();
                    queue.addContainerTransaction(new ContainerTransaction(p.getName(), (byte)1, container,
                            cursor.getType().name(), amt, loc.getWorld().getName(),
                            loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
                }
            }
            case MOVE_TO_OTHER_INVENTORY -> {
                if (current != null && current.getType() != Material.AIR) {
                    byte dir = e.getInventory().equals(clicked) ? (byte)0 : (byte)1;
                    queue.addContainerTransaction(new ContainerTransaction(p.getName(), dir, container,
                            current.getType().name(), current.getAmount(), loc.getWorld().getName(),
                            loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onDrag(InventoryDragEvent e) {
        if (!(e.getWhoClicked() instanceof org.bukkit.entity.Player p)) return;
        if (!isContainer(e.getInventory().getType())) return;
        var loc = getContainerLocation(e.getInventory());
        if (loc == null) return;

        var dragged = e.getOldCursor();
        if (dragged == null || dragged.getType() == Material.AIR) return;

        int total = e.getNewItems().values().stream()
                .mapToInt(ItemStack::getAmount)
                .sum();

        if (total > 0) {
            queue.addContainerTransaction(new ContainerTransaction(p.getName(), (byte)1,
                    e.getInventory().getType().name(), dragged.getType().name(), total,
                    loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
        }
    }

    private boolean isContainer(InventoryType t) {
        return switch (t) {
            case CHEST, ENDER_CHEST, BARREL, SHULKER_BOX, HOPPER, DROPPER,
                 DISPENSER, FURNACE, BLAST_FURNACE, SMOKER, BREWING,
                 ANVIL, BEACON, ENCHANTING, MERCHANT, CARTOGRAPHY,
                 GRINDSTONE, LECTERN, LOOM, STONECUTTER, SMITHING -> true;
            default -> false;
        };
    }

    private Location getContainerLocation(Inventory i) {
        var h = i.getHolder();
        if (h instanceof BlockState bs) return bs.getLocation();
        if (h instanceof DoubleChest dc) return dc.getLocation();
        return null;
    }
}

File: src/main/kotlin/cat/daisy/core/events/ContainerListeners.kt
package cat.daisy.core.events

import cat.daisy.core.service.ActionLogService
import cat.daisy.core.service.ConfigService
import org.bukkit.Material
import org.bukkit.block.BlockState
import org.bukkit.block.DoubleChest
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryAction
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryDragEvent
import org.bukkit.event.inventory.InventoryType
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack

class ContainerListeners(
    private val logs: ActionLogService,
) : Listener {
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onClick(e: InventoryClickEvent) {
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "enabled", true)) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.container", true)) return
        val p = e.whoClicked as? org.bukkit.entity.Player ?: return
        val clicked = e.clickedInventory ?: return
        if (!isContainer(clicked.type)) return
        val loc = getContainerLocation(clicked) ?: return

        val current = e.currentItem
        val cursor = e.cursor
        val container = clicked.type.name

        when (e.action) {
            InventoryAction.PICKUP_ALL,
            InventoryAction.PICKUP_SOME,
            InventoryAction.PICKUP_HALF,
            InventoryAction.PICKUP_ONE,
            -> {
                if (current != null && current.type != Material.AIR) {
                    val amt =
                        when (e.action) {
                            InventoryAction.PICKUP_HALF -> (current.amount + 1) / 2
                            InventoryAction.PICKUP_ONE -> 1
                            else -> current.amount
                        }
                    logs.submitContainer(
                        ActionLogService.ContainerTx(
                            time = (System.currentTimeMillis() / 1000L).toInt(),
                            playerName = p.name,
                            action = 0, // pickup/remove
                            containerType = container,
                            material = current.type.name,
                            amount = amt,
                            world = loc.world!!.name,
                            x = loc.blockX,
                            y = loc.blockY,
                            z = loc.blockZ,
                        ),
                    )
                }
            }

            InventoryAction.PLACE_ALL,
            InventoryAction.PLACE_SOME,
            InventoryAction.PLACE_ONE,
            -> {
                if (cursor != null && cursor.type != Material.AIR) {
                    val amt = if (e.action == InventoryAction.PLACE_ONE) 1 else cursor.amount
                    logs.submitContainer(
                        ActionLogService.ContainerTx(
                            time = (System.currentTimeMillis() / 1000L).toInt(),
                            playerName = p.name,
                            action = 1, // place/insert
                            containerType = container,
                            material = cursor.type.name,
                            amount = amt,
                            world = loc.world!!.name,
                            x = loc.blockX,
                            y = loc.blockY,
                            z = loc.blockZ,
                        ),
                    )
                }
            }

            InventoryAction.MOVE_TO_OTHER_INVENTORY -> {
                if (current != null && current.type != Material.AIR) {
                    val dir: Int = if (e.inventory == clicked) 0 else 1
                    logs.submitContainer(
                        ActionLogService.ContainerTx(
                            time = (System.currentTimeMillis() / 1000L).toInt(),
                            playerName = p.name,
                            action = dir,
                            containerType = container,
                            material = current.type.name,
                            amount = current.amount,
                            world = loc.world!!.name,
                            x = loc.blockX,
                            y = loc.blockY,
                            z = loc.blockZ,
                        ),
                    )
                }
            }

            else -> {}
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onDrag(e: InventoryDragEvent) {
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "enabled", true)) return
        if (!ConfigService.getModuleSettingBoolean(ConfigService.EVENTS, "actionlogger", "capture.container", true)) return
        val p = e.whoClicked as? org.bukkit.entity.Player ?: return
        if (!isContainer(e.inventory.type)) return
        val loc = getContainerLocation(e.inventory) ?: return

        val dragged = e.oldCursor ?: return
        if (dragged.type == Material.AIR) return

        val total = e.newItems.values.sumOf(ItemStack::getAmount)
        if (total > 0) {
            logs.submitContainer(
                ActionLogService.ContainerTx(
                    time = (System.currentTimeMillis() / 1000L).toInt(),
                    playerName = p.name,
                    action = 1,
                    containerType = e.inventory.type.name,
                    material = dragged.type.name,
                    amount = total,
                    world = loc.world!!.name,
                    x = loc.blockX,
                    y = loc.blockY,
                    z = loc.blockZ,
                ),
            )
        }
    }

    private fun isContainer(t: InventoryType): Boolean =
        when (t) {
            InventoryType.CHEST,
            InventoryType.ENDER_CHEST,
            InventoryType.BARREL,
            InventoryType.SHULKER_BOX,
            InventoryType.HOPPER,
            InventoryType.DROPPER,
            InventoryType.DISPENSER,
            InventoryType.FURNACE,
            InventoryType.BLAST_FURNACE,
            InventoryType.SMOKER,
            InventoryType.BREWING,
            InventoryType.ANVIL,
            InventoryType.BEACON,
            InventoryType.ENCHANTING,
            InventoryType.MERCHANT,
            InventoryType.CARTOGRAPHY,
            InventoryType.GRINDSTONE,
            InventoryType.LECTERN,
            InventoryType.LOOM,
            InventoryType.STONECUTTER,
            InventoryType.SMITHING,
            -> true

            else -> false
        }

    private fun getContainerLocation(i: Inventory) =
        when (val h = i.holder) {
            is BlockState -> h.location
            is DoubleChest -> h.location
            else -> null
        }
}
